core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,395 @@
+import org.infinispan.util.concurrent.TimeoutException;
+import org.infinispan.util.concurrent.locks.DeadlockChecker;
+import org.infinispan.util.concurrent.locks.DeadlockDetectedException;
+import org.infinispan.util.concurrent.locks.ExtendedLockPromise;
+import org.infinispan.util.concurrent.locks.KeyAwareLockListener;
+import org.infinispan.util.concurrent.locks.KeyAwareLockPromise;
+import org.infinispan.util.concurrent.locks.LockListener;
+import org.infinispan.util.concurrent.locks.LockManager;
+import org.infinispan.util.concurrent.locks.LockPromise;
+import org.infinispan.util.concurrent.locks.LockState;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+/**
+ * The default {@link LockManager} implementation for transactional and non-transactional caches.
+ *
+ * @author Pedro Ruivo
+ * @since 8.0
+ */
+@MBean(objectName = "LockManager", description = "Manager that handles MVCC locks for entries")
+public class DefaultLockManager implements LockManager {
+
+   private static final Log log = LogFactory.getLog(DefaultLockManager.class);
+   private static final boolean trace = log.isTraceEnabled();
+   private static final AtomicReferenceFieldUpdater<CompositeLockPromise, LockState> UPDATER =
+         newUpdater(CompositeLockPromise.class, LockState.class, "lockState");
+
+
+   protected LockContainer lockContainer;
+   protected Configuration configuration;
+   protected ScheduledExecutorService scheduler;
+
+   @Inject
+   public void inject(LockContainer container, Configuration configuration,
+                      @ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService executorService) {
+      this.lockContainer = container;
+      this.configuration = configuration;
+      this.scheduler = executorService;
+   }
+
+
+   @Override
+   public KeyAwareLockPromise lock(Object key, Object lockOwner, long time, TimeUnit unit) {
+      Objects.requireNonNull(key, "Key must be non null");
+      Objects.requireNonNull(lockOwner, "Lock owner must be non null");
+      Objects.requireNonNull(unit, "Time unit must be non null");
+
+      if (trace) {
+         log.tracef("Lock key=%s for owner=%s. timeout=%s (%s)", key, lockOwner, time, unit);
+      }
+
+      ExtendedLockPromise promise = lockContainer.acquire(key, lockOwner, time, unit);
+      return new KeyAwareExtendedLockPromise(promise, key, unit.toMillis(time)).scheduleLockTimeoutTask(scheduler);
```

Comment:
I think about it but there is one use case where I couldn't fit it. 
In extended-stats module, if you have a tx acquiring k1, k2 and striped locking configured and if k1 and k2 will hash on the same InfinispanLock, I would need a way to trigger 2 notifications with 2 different keys.